gracefully shutdown reporter & add agent-test-tool based e2e case

tisonkun opened the pull request:
Signed-off-by: tison <wander4096@gmail.com>

cc @wu-sheng @Shikugawa

tisonkun committed:
self-contained example & gracefully shutdown reporter

Signed-off-by: tison <wander4096@gmail.com>

tisonkun committed:
tokio-stream can be a dev-dependencies

Signed-off-by: tison <wander4096@gmail.com>

tisonkun commented:
I noticed that we may not use mpsc::channel at all.

Currently, the reporter flushes per context on the streaming rpc. May you provide other languages' reporter implementation as references and I can work for a proper API.

wu-sheng commented:
Hi, if you want self-contained, this is another example. As a client, the official way to test is set up e2e, which this lib is only working as a client.
Self-contained is tricky because without SkyWalking OAP running, this client could send tracing w/ grpc means nothing.

wu-sheng commented:
This is how the test thing works, https://github.com/SkyAPM/cpp2sky/blob/main/.github/workflows/main.yml

wu-sheng commented:
I would like to recommend you to add an e2e test case for this repo, which has more value from our experiences.

Shikugawa commented:
@tisonkun Thank you for awesome contribution! If you are interested in e2e framework, you should see [here](https://github.com/Shikugawa/rs2sky/tree/main/tests/e2e) (This is my experimental work of e2e, but it is not ready for production). It may help you.

Shikugawa commented on src/reporter/grpc.rs:
```diff
@@ -51,23 +53,51 @@ impl Reporter {
+                }
+            }
+            rx.close();
             while let Some(message) = rx.recv().await {
```
It may not be required anymore.

tisonkun replied:
No. If a select chooses shutdown first, there can be outstanding messages to be processed. We close the `rx` so that there is no more inflight message, but the outstanding ones should be processed.

See also https://docs.rs/tokio/1.15.0/tokio/sync/mpsc/struct.Receiver.html#method.close for example.

Shikugawa commented on examples/simple_trace_report.rs:
```diff
@@ -1,14 +1,67 @@
+use std::error::Error;
```
Could we add new file which includes self-contained example?

tisonkun replied:
Yes. Let me try to build an e2e test for it.

Shikugawa reviewed (changes requested):
Sorry, I'm not sure how situation we should use self-contained reporters. Could you show me the usecase for it?

tisonkun commented:
> I would like to recommend you to add an e2e test case for this repo, which has more value from our experiences.

That makes sense. I'm confused about how to run the example properly and failed to find a resource to set up the environment.

Thanks for your information. Let me dig it out.

wu-sheng commented:
> That makes sense. I'm confused about how to run the example properly and failed to find a resource to set up the environment.

Because you need a backend and UI, https://skywalking.apache.org/docs/main/v8.9.1/readme/. 
About the self-hosted, it is not 100% meaningless, but mostly, it just could print the segment values.

Shikugawa commented:
> > I would like to recommend you to add an e2e test case for this repo, which has more value from our experiences.
> 
> That makes sense. I'm confused about how to run the example properly and failed to find a resource to set up the environment.
> 
> Thanks for your information. Let me dig it out.

We preapred simple docker-compose envs [here](https://github.com/apache/skywalking-rust/blob/master/docker-compose.e2e.yml).
Also, your change will break e2e code [here](https://github.com/apache/skywalking-rust/blob/master/tests/e2e/src/main.rs).

tisonkun commented:
@wu-sheng It seems the example can leave as is and I can verify the graceful shutdown logic by fixing [current e2e test](https://github.com/apache/skywalking-rust/blob/master/tests/e2e/src/main.rs).

wu-sheng commented:
@tisonkun E2E has a very specific meaning in the SkyWalking context. 

There are 2 kinds of E2E tests.
1. E2E with SkyWalking backend. (I think this isn't @Shikugawa 's suggestion, just mention it here to make the context clear)
You could learn [this tool](https://skywalking.apache.org/docs/skywalking-infra-e2e/latest/readme/), which is an established sub-project for a docker-compose/KinD(k8s) env testing to verify how components work with each other because the SkyWalking ecosystem is super large and complex.

2. Plugin tests with [agent-test-tool](https://github.com/apache/skywalking-agent-test-tool), which is a mock OAP server to collect and print data to a YAML for you. Here is [how it is used in cpp2sky](https://github.com/SkyAPM/cpp2sky/blob/34053a05c3e1b6304704f074f648238dac733564/docker-compose.e2e.yml?_pjax=%23js-repo-pjax-container%3Afirst-of-type%2C%20div%5Bitemtype%3D%22http%3A%2F%2Fschema.org%2FSoftwareSourceCode%22%5D%20main%3Afirst-of-type%2C%20%5Bdata-pjax-container%5D%3Afirst-of-type#L6).

___
You should follow (2) more for the rust plugin. This test is working as `rust-a` -> `rust-b`, and both of them are sending trace data to `mock-collector`. Then verifying the data from `mock-collector` with the expected data file you provided.

tisonkun commented:
@wu-sheng @Shikugawa I can try to fix and verify e2e failure locally, but it seems our CI doesn't check the e2e test. Shall we add it also?

wu-sheng commented:
Yes, e2e is a new case.

tisonkun committed:
revert example should not start a server

Signed-off-by: tison <wander4096@gmail.com>

tisonkun committed:
adapt e2e test logic

Signed-off-by: tison <wander4096@gmail.com>

tisonkun committed:
enable e2e test

Signed-off-by: tison <wander4096@gmail.com>

tisonkun commented:
@wu-sheng the current tests under `tests/e2e` folder is of type (2). So I'm trying to enable this test to proceeding my patch.

I've verified it locally by:

```
docker run -it -p 19876:19876 -p 12800:12800 skywalking-rust_collector
cargo run -- --mode consumer # under tests/e2e/
cargo run -- --mode producer # under tests/e2e/
python3 tests/e2e/run_e2e.py --expected_file=tests/e2e/data/expected_context.yaml  --max_retry_times=3 --target_path=/ping # under root path
```

but fighting with bring it to docker-compose based e2e config...

cc @Shikugawa

tisonkun commented:
It seems that `depends_on` doesn't take effect and consumer/producer runs before collector and panic...I don't know how to process then. If I manually start service parts one by one, the test can pass.

tisonkun committed:
add healthcheck

Signed-off-by: tison <wander4096@gmail.com>

kezhenxu94 commented:
> It seems that `depends_on` doesn't take effect and consumer/producer runs before collector and panic...I don't know how to process then. If I manually start service parts one by one, the test can pass.

depends_on should take effects, what is out of expectation is that consumer/provider tris to do something that OAP is not yet ready. Please use 

```yaml
depends_on:
  oap:
    condition: service_healthy
```

And add a health check for OAP. 

NOTE: version 3.7 of docker compose file might not accept this syntax but I know 3.8 and 2.1 can.

wu-sheng commented:
I have seen you added a health check. This is important.

tisonkun committed:
e2e main supports healthCheck

Signed-off-by: tison <wander4096@gmail.com>

tisonkun committed:
reconcil docker-compose file

Signed-off-by: tison <wander4096@gmail.com>

tisonkun committed:
revrt e2e main healthCheck

Adding healthCheck causes hang without explicit reason.

Given that the consumer started quickly, I tend to revert the healthCheck for now.

Signed-off-by: tison <wander4096@gmail.com>

tisonkun commented:
Finally I hack through it.

I think this is ready for review, but I still confuse why https://github.com/apache/skywalking-rust/pull/9/commits/bbda7cfc539dfecf71254f2db0ad16fe51caa1c7 causes hanging forever :(

tisonkun committed:
e2e main healthCheck using image with curl

Signed-off-by: tison <wander4096@gmail.com>

wu-sheng commented on .github/workflows/e2e.yml:
```diff
@@ -0,0 +1,25 @@
+          pip3 install --upgrade pip
+          pip3 install setuptools
+          pip3 install -r requirements.txt
+          python3 tests/e2e/run_e2e.py --expected_file=tests/e2e/data/expected_context.yaml --max_retry_times=3 --target_path=/ping
```
I can't find the `tests/e2e/data/expected_context.yaml`. How does this test pass? Could you check the logs?
Is this a `run_e2e.py` bug or something?

tisonkun replied:
It's already in the codebase here: https://github.com/apache/skywalking-rust/blob/1465d9fc6811dcb009e75f5c1dd2d5b1ef7a52a9/tests/e2e/data/expected_context.yaml

I don't add a new test case but according to your suggestion https://github.com/apache/skywalking-rust/pull/9#issuecomment-1008699575, enable the existing but not in regression e2e test - it's a plugin test based on agent-test-tool.

wu-sheng commented:
@tisonkun Please rename the title to make the changes clear.

tisonkun commented:
@wu-sheng updated. Notify me if it's better to split into two PRs.

wu-sheng commented:
I am fine that these two are in one PR, as this repo is a very early stage.

But I hope you could recheck the test coverage
> src/context/trace_context.rs | 69.81% <0.00%> (-6.48%)

I am not sure whether e2e testing could be counted in Rust, but it can be in the Java.

tisonkun commented:
@wu-sheng I checked the report and don't think it's a real regression:

1. "Regressions" on `impl fmt:Debug`. It's meaningless to test an standard lib implementation.
2. "Regressions" on `shutdown`/`sender`.  The e2e test should cover it.

Shikugawa reviewed (approved):
Great. Thank you for changing!

wu-sheng commented:
I got a notification that e2e of merged commit has failed in 9s after e2e began. 
It makes me feel the health check seems not working well.
I rerun the test, and will see whether it could pass.

tisonkun commented:
@wu-sheng @Shikugawa according to the failure report, it seems the received data is unstable (not stably matches the expected data).

https://github.com/apache/skywalking-rust/runs/4818192129?check_suite_focus=true

```
diff list: 
- meterItems: []
  segmentItems:
  - segmentSize: 1
    segments:
-   - segmentId: '184682932856352416001309273098852514966'
+   - segmentId: not null
      spans:
      - componentId: 11000
-       endTime: 1642171583
+       endTime: gt 0
        isError: false
+       operationId: 0
        operationName: /pong
        parentSpanId: 1
        peer: consumer:8082
        skipAnalysis: false
        spanId: 2
        spanLayer: Http
        spanType: Exit
-       startTime: 1642171583
+       startTime: gt 0
      - componentId: 11000
-       endTime: 1642171583
+       endTime: gt 0
        isError: false
+       operationId: 0
        operationName: /ping
        parentSpanId: 0
        peer: ''
        skipAnalysis: false
        spanId: 1
        spanLayer: Http
        spanType: Entry
-       startTime: 1642171583
+       startTime: gt 0
    serviceName: producer
  - segmentSize: 1
    segments:
-   - segmentId: '251652345953849008728622634953354684403'
+   - segmentId: not null
      spans:
      - componentId: 11000
-       endTime: 1642171583
+       endTime: gt 0
        isError: false
+       operationId: 0
        operationName: /pong
        parentSpanId: 0
        peer: ''
        refs:
        - networkAddress: consumer:8082
          parentEndpoint: /pong
          parentService: producer
          parentServiceInstance: node_0
          parentSpanId: 2
-         parentTraceSegmentId: '184682932856352416001309273098852514966'
+         parentTraceSegmentId: not null
          refType: CrossProcess
-         traceId: '280108941388649570187429178111613482643'
+         traceId: not null
        skipAnalysis: false
        spanId: 1
        spanLayer: Http
        spanType: Entry
-       startTime: 1642171583
+       startTime: gt 0
    serviceName: consumer
Traceback (most recent call last):
  File "tests/e2e/run_e2e.py", line 60, in <module>
    assert res.status_code == 200
AssertionError
```

wu-sheng commented:
I think the key point is `AssertionError`, which seems the observed services are not available.
